src/toil/worker.py

```diff
@@ -54,29 +55,102 @@ class StatsDict(MagicExpando):
 
     jobs: List[str]
 
+def checkSuccessorReadyToRunMultiplePredecessors(
+    successor: JobDescription,
+    predecessor: JobDescription,
+    jobStore: AbstractJobStore,
+    toilState: ToilState,
+) -> bool:
+    """
+    Handle the special cases of checking if a successor job is
+    ready to run when there are multiple predecessors.
+
+    :param successor: The successor which has failed.
+    :param predecessor: The job which the successor comes after.
+
+    """
+    # See implementation note at the top of this file for discussion of multiple predecessors
+    logger.debug(
+        "Successor job: %s of job: %s has multiple " "predecessors",
+        successor,
+        predecessor,
+    )
+    logger.debug(
+        "Already finished predecessors are: %s", successor.predecessorsFinished
+    )
+
+    # Get the successor JobDescription, which is cached
+    if successor.jobStoreID not in toilState.jobsToBeScheduledWithMultiplePredecessors:
+        # TODO: We're loading from the job store in an ad-hoc way!
+        loaded = jobStore.load(successor.jobStoreID)
+        toilState.jobsToBeScheduledWithMultiplePredecessors[
+            successor.jobStoreID
+        ] = loaded
+    # TODO: we're clobbering a JobDescription we're passing around by value.
+    successor = toilState.jobsToBeScheduledWithMultiplePredecessors[
+        successor.jobStoreID
+    ]
```

Comment:
Hey @douglowe , it has been clarified that `jobsToBeScheduledWithMultiplePredecessors` is a `set[str]`, not a dictionary. Could you refresh this test in light of changes made since you started this PR?

Thanks again!

Reply:
Ahh - yes, I will correct that now!

Reply:
Right - my brain is slowly getting back up to speed with this. Not quite the quick fix I was thinking of - as my code was using the dictionary of job information cached here to check what successors the current job had.

I think that a discussion will be needed with the Toil devs, to see where the main Toil development arc for the JobStore is heading. IIRC the intent was to have a mutable central JobStore database, which could be accessed & changed by all jobs (not just the parent job) - which would make the hashed jobstore that my hacky code passes around redundant.

Perhaps we should cut out the test code that I wrote (though this also needs some work, as the regex doesn't match the current output stored in the Stats directory), and save that - as it still shows what kind of workflow in which I'd hope full job chaining would occur. Then, separately, we can work on correcting the information flow and logic for working out what jobs can be chained and which can't?